File: parser/src/parser.rs
use crate::lexer::{Token, Tokens};
use line_col::LineColLookup;
use std::ops::Range;

pub type Result<T> = std::result::Result<T, Error>;

#[derive(Debug)]
pub struct Error {
    pub message: String,
    pub location: Option<SourceLocation>,
}

impl Error {
    fn new(message: &str, offset: Option<&Range<usize>>, lc: &LineColLookup) -> Self {
        Error {
            message: String::from(message),
            location: offset.map(|offset| SourceLocation::new(offset, lc)),
        }
    }
}

#[derive(Debug)]
pub struct Ast {
    pub node: AstNode,
    pub location: SourceLocation,
}

impl Ast {
    fn new(node: AstNode, offset: &Range<usize>, lc: &LineColLookup) -> Self {
        Ast {
            node,
            location: SourceLocation::new(offset, lc),
        }
    }
}

#[derive(Debug)]
pub enum AstNode {
    Block(Vec<AstNodeStatement>),
    List(Vec<Ast>),
    Name(String),
    Text(String),
    Number(String),
    Literal(Box<Ast>),
    Escaped(Box<Ast>),
}

#[derive(Debug)]
pub struct AstNodeStatement {
    pub items: Vec<Ast>,
    pub location: SourceLocation,
}

#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub struct SourceLocation {
    pub line: usize,
    pub column: usize,
}

macro_rules! parse_single_token {
    ($token:ident, $tokens:expr) => {
        Ok(match $tokens.peek() {
            Some((Token::$token, _)) => {
                $tokens.next();
                Some(())
            }
            _ => None,
        })
    };

    ($token:ident, $tokens:expr, $lc:expr) => {
        Ok(match $tokens.peek() {
            Some((Token::$token(value), offset)) => {
                $tokens.next();
                Some(Ast::new(AstNode::$token(value.clone()), offset, $lc))
            }
            _ => None,
        })
    };
}

macro_rules! parse_prefix_token {
    ($token:ident => $node:ident, $tokens:expr, $lc:expr) => {{
        let offset = match $tokens.peek() {
            Some((Token::$token, offset)) => {
                $tokens.next();
                offset
            }
            _ => return Ok(None),
        };

        let value =
            parse_value($tokens, $lc)?.ok_or_else(|| Error::new("Expected value", None, $lc))?;

        Ok(Some(Ast::new(AstNode::$node(Box::new(value)), offset, $lc)))
    }};
}

pub fn parse_block(tokens: &mut Tokens, lc: &LineColLookup) -> Result<Option<Ast>> {
    let start_offset = match tokens.peek() {
        Some((Token::OpenBrace, offset)) => {
            tokens.next();
            offset
        }
        _ => return Ok(None),
    };

    let statements = parse_statements(tokens, lc)?;

    parse_newlines(tokens)?;

    if let Some((token, end_offset)) = tokens.next() {
        match token {
            Token::CloseBrace => Ok(Some(Ast::new(AstNode::Block(statements), start_offset, lc))),
            _ => Err(Error::new("Expected '}'", Some(end_offset), lc)),
        }
    } else {
        Err(Error::new("Expected '}'", None, lc))
    }
}

pub fn parse_file(tokens: &mut Tokens, lc: &LineColLookup) -> Result<Ast> {
    parse_single_token!(Shebang, tokens)?;

    let statements = parse_statements(tokens, lc)?;

    parse_newlines(tokens)?;

    match tokens.next() {
        None => Ok(Ast::new(AstNode::Block(statements), &(0..0), lc)),
        Some((_, offset)) => Err(Error::new("Unexpected token", Some(offset), lc)),
    }
}

pub fn parse_inline_program(tokens: &mut Tokens, lc: &LineColLookup) -> Result<Ast> {
    parse_single_token!(Shebang, tokens)?;

    let mut values = Vec::new();

    while let Some(value) = parse_value(tokens, lc)? {
        values.push(value);
    }

    match tokens.next() {
        None => Ok(Ast::new(AstNode::List(values), &(0..0), lc)),
        Some((_, offset)) => Err(Error::new("Unexpected token", Some(offset), lc)),
    }
}

fn parse_statements(tokens: &mut Tokens, lc: &LineColLookup) -> Result<Vec<AstNodeStatement>> {
    let mut values: Vec<(Vec<Ast>, Option<SourceLocation>)> = vec![(Vec::new(), None)];

    loop {
        if parse_newline(tokens)?.is_some() {
            values.push((Vec::new(), None));
            continue;
        }

        parse_newlines(tokens)?;

        if let Some(value) = parse_value(tokens, lc)? {
            let (items, location) = values.last_mut().unwrap();
            location.get_or_insert(value.location);
            items.push(value);
        } else {
            break;
        }
    }

    let statements = values
        .into_iter()
        .filter_map(|(items, location)| {
            location.map(|location| AstNodeStatement { items, location })
        })
        .collect::<Vec<_>>();

    Ok(statements)
}

pub fn parse_list(tokens: &mut Tokens, lc: &LineColLookup) -> Result<Option<Ast>> {
    let start_offset = match tokens.peek() {
        Some((Token::OpenParenthesis, offset)) => {
            tokens.next();
            offset
        }
        _ => return Ok(None),
    };

    let mut values = Vec::new();

    loop {
        parse_newlines(tokens)?;

        if let Some(value) = parse_value(tokens, lc)? {
            values.push(value);
        } else {
            break;
        }
    }

    parse_newlines(tokens)?;

    if let Some((token, end_offset)) = tokens.next() {
        match token {
            Token::CloseParenthesis => Ok(Some(Ast::new(AstNode::List(values), start_offset, lc))),
            _ => Err(Error::new("Expected ')'", Some(end_offset), lc)),
        }
    } else {
        Err(Error::new("Expected ')'", None, lc))
    }
}

pub fn parse_brackets(tokens: &mut Tokens, lc: &LineColLookup) -> Result<Option<Ast>> {
    match tokens.peek() {
        Some((Token::OpenBracket, offset)) => {
            Err(Error::new("Bracket syntax is reserved", Some(offset), lc))
        }
        _ => Ok(None),
    }
}

pub fn parse_name(tokens: &mut Tokens, lc: &LineColLookup) -> Result<Option<Ast>> {
    parse_single_token!(Name, tokens, lc)
}

pub fn parse_text(tokens: &mut Tokens, lc: &LineColLookup) -> Result<Option<Ast>> {
    parse_single_token!(Text, tokens, lc)
}

pub fn parse_number(tokens: &mut Tokens, lc: &LineColLookup) -> Result<Option<Ast>> {
    parse_single_token!(Number, tokens, lc)
}

pub fn parse_literal(tokens: &mut Tokens, lc: &LineColLookup) -> Result<Option<Ast>> {
    parse_prefix_token!(Quote => Literal, tokens, lc)
}

pub fn parse_escaped(tokens: &mut Tokens, lc: &LineColLookup) -> Result<Option<Ast>> {
    parse_prefix_token!(Backslash => Escaped, tokens, lc)
}

fn parse_value(tokens: &mut Tokens, lc: &LineColLookup) -> Result<Option<Ast>> {
    let choices = &[
        parse_block,
        parse_list,
        parse_brackets,
        parse_name,
        parse_text,
        parse_number,
        parse_literal,
        parse_escaped,
    ];

    for choice in choices {
        if let Some(value) = choice(tokens, lc)? {
            return Ok(Some(value));
        }
    }

    Ok(None)
}

fn parse_newline(tokens: &mut Tokens) -> Result<Option<()>> {
    parse_single_token!(Newline, tokens)
}

fn parse_newlines(tokens: &mut Tokens) -> Result<()> {
    while parse_newline(tokens)?.is_some() {}

    Ok(())
}

impl SourceLocation {
    pub fn new(offset: &Range<usize>, lc: &LineColLookup) -> Self {
        let (line, column) = lc.get(offset.start);
        SourceLocation { line, column }
    }
}

File: Parser/Sources/Parser.swift
import Foundation
import Wipple

public struct ParseError: Swift.Error {
    public let message: String
    public let location: SourceLocation?

    public init(_ message: String, _ location: SourceLocation? = nil) {
        self.message = message
        self.location = location
    }
}

public struct AST: Hashable {
    public var node: Node
    public var location: SourceLocation

    public init(_ node: Node, _ location: SourceLocation) {
        self.node = node
        self.location = location
    }

    public indirect enum Node: Hashable {
        case block([Statement])
        case list([AST])
        case name(String)
        case number(Decimal)
        case text(String)
        case literal(AST)
        case escaped(AST)

        public struct Statement: Hashable {
            public var items: [AST]
            public var location: SourceLocation
        }
    }
}

public typealias TokenIterator = _PeekableIterator<Array<TokenWithLocation>.Iterator>

public extension Array where Element == TokenWithLocation {
    func forParsing() -> TokenIterator {
        _PeekableIterator(
            lazy
                .filter { token in
                    switch token.token {
                    case .whitespace, .comment:
                        return false
                    default:
                        return true
                    }
                }
                .makeIterator()
        )
    }
}

public func parseBlock(_ tokens: TokenIterator) throws -> AST? {
    guard case let (.openBrace, start)? = tokens.peek()?.tuple else {
        return nil
    }

    tokens.next()

    let statements = try parseStatements(tokens)

    parseNewlines(tokens)

    if let (token, end) = tokens.next()?.tuple {
        guard case .closeBrace = token else {
            throw ParseError("Expected '}'", end)
        }

        return AST(.block(statements), start)
    }

    throw ParseError("Expected '}'")
}

public func parseFile(_ tokens: TokenIterator) throws -> AST {
    let start: SourceLocation
    if let (_, location) = tokens.peek()?.tuple {
        start = SourceLocation(file: location.file, line: 1, column: 1)
    } else {
        start = SourceLocation(line: 1, column: 1)
    }

    let statements = try parseStatements(tokens)

    parseNewlines(tokens)

    if let (_, location) = tokens.next()?.tuple {
        throw ParseError("Unexpected token", location)
    }

    return AST(.block(statements), start)
}

public func parseInlineProgram(_ tokens: TokenIterator) throws -> AST {
    let start: SourceLocation
    if let (_, location) = tokens.peek()?.tuple {
        start = SourceLocation(file: location.file, line: 1, column: 1)
    } else {
        start = SourceLocation(line: 1, column: 1)
    }

    var values: [AST] = []

    while let value = try parseValue(tokens) {
        values.append(value)
    }

    if let (_, location) = tokens.next()?.tuple {
        throw ParseError("Unexpected token", location)
    }

    return AST(.list(values), start)
}

private func parseStatements(_ tokens: TokenIterator) throws -> [AST.Node.Statement] {
    var statements: [(items: [AST], location: SourceLocation?)] = [([], nil)]

    while true {
        if parseNewline(tokens) {
            statements.append(([], nil))
            continue
        }

        parseNewlines(tokens)

        guard let value = try parseValue(tokens) else {
            break
        }

        let index = statements.endIndex - 1

        statements[index].items.append(value)

        if statements[index].location == nil {
            statements[index].location = value.location
        }
    }

    return statements.compactMap {
        let (items, location) = $0
        return location.map { .init(items: items, location: $0) }
    }
}

public func parseList(_ tokens: TokenIterator) throws -> AST? {
    guard case let (.openParenthesis, start)? = tokens.peek()?.tuple else {
        return nil
    }

    tokens.next()

    var items: [AST] = []

    while true {
        parseNewlines(tokens)

        guard let value = try parseValue(tokens) else {
            break
        }

        items.append(value)
    }

    if let (token, end) = tokens.next()?.tuple {
        guard case .closeParenthesis = token else {
            throw ParseError("Expected ')'", end)
        }

        return AST(.list(items), start)
    }

    throw ParseError("Expected ')'")
}

public func parseBrackets(_ tokens: TokenIterator) throws -> AST? {
    if case let (.openBracket, start)? = tokens.peek()?.tuple {
        throw ParseError("Bracket syntax is reserved", start)
    }

    return nil
}

public func parseName(_ tokens: TokenIterator) throws -> AST? {
    guard case let (.name(name), start)? = tokens.peek()?.tuple else {
        return nil
    }

    tokens.next()

    return AST(.name(name), start)
}

public func parseNumber(_ tokens: TokenIterator) throws -> AST? {
    guard case let (.number(number), start)? = tokens.peek()?.tuple else {
        return nil
    }

    tokens.next()

    return AST(.number(number), start)
}

public func parseText(_ tokens: TokenIterator) throws -> AST? {
    guard case let (.text(text), start)? = tokens.peek()?.tuple else {
        return nil
    }

    tokens.next()

    return AST(.text(text), start)
}

public func parseLiteral(_ tokens: TokenIterator) throws -> AST? {
    guard case let (.quote, start)? = tokens.peek()?.tuple else {
        return nil
    }

    tokens.next()

    guard let value = try parseValue(tokens) else {
        throw ParseError("Expected value", start)
    }

    return AST(.literal(value), start)
}

public func parseEscaped(_ tokens: TokenIterator) throws -> AST? {
    guard case let (.backslash, start)? = tokens.peek()?.tuple else {
        return nil
    }

    tokens.next()

    guard let value = try parseValue(tokens) else {
        throw ParseError("Expected value", start)
    }

    return AST(.escaped(value), start)
}

private func parseValue(_ tokens: TokenIterator) throws -> AST? {
    let choices = [
        parseBlock,
        parseList,
        parseBrackets,
        parseName,
        parseText,
        parseNumber,
        parseLiteral,
        parseEscaped,
    ]

    for choice in choices {
        if let value = try choice(tokens) {
            return value
        }
    }

    return nil
}

private func parseNewline(_ tokens: TokenIterator) -> Bool {
    guard case (.newline, _)? = tokens.peek()?.tuple else {
        return false
    }

    tokens.next()

    return true
}

private func parseNewlines(_ tokens: TokenIterator) {
    while parseNewline(tokens) {}
}

public class _PeekableIterator<Base: IteratorProtocol>: IteratorProtocol {
    private var peeked: Base.Element?
    private var iter: Base

    public init(_ base: Base) {
        iter = base
        peeked = iter.next()
    }

    public func peek() -> Base.Element? {
        peeked
    }

    @discardableResult
    public func next() -> Base.Element? {
        let result = peeked

        if peeked != nil {
            peeked = iter.next()
        }

        return result
    }
}
